### PR TITLE
Easier Command Based programming

### PR DESCRIPTION
**Pull request by FRC2539**
This adds four classes to make writing command based programs easier in Python. We've also included an example program that uses these classes. If this is accepted, we'd be willing to make a page for the readthedocs.org site explaining how it can be used.

All four helper classes are in a package called `commandbased`:

- `CommandBasedRobot` serves as a higher level `IterativeRobot` and provides default implementations of all periodic functions, along with exception handling for competitions
- `Command` provides sensible defaults for WPILib's `Command` class, so that we only need to implement the unique parts of each command.
- `TimedCommand` implements a command that runs for a set period of time before ending automatically.
- `InstantCommand` implements a command that runs only the `initialize` method and then ends.

### CODE REVIEW

File: commandbased/commandbasedrobot.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    def __runPeriodic(self, periodicFunction):
+        '''
+        Run the correct "periodic" function and then wait for new data from the
+        driver station.
+        '''
+
+        periodicFunction()
+
+        while not self.ds.isNewControlData():
```
Comment (virtuald) [inline commandbased/commandbasedrobot.py]: This seems suspect to me. Does waitForData not work?

File: commandbased/timedcommand.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+class TimedCommand(Command):
+    '''A command that runs for a set period of time.'''
+
+    def __init__(self, name, timeoutInSeconds):
+        super().__init__(name, timeoutInSeconds)
+
+
+    def isFinished(self):
+        return self.isTimedOut()
```
Comment (virtuald) [inline commandbased/timedcommand.py]: No default implementation to hint to the user that they should override it?

Comment (virtuald) [inline commandbased/timedcommand.py]: Never mind, that shows how much I don't know about commands.

**Comment by virtuald** [comment]
You should definitely submit TimedCommand and Instance command as a PR to wpilib, they make a lot of sense.

**Comment by virtuald** [comment]
Other than the waitForData comment, this all seems fine to me and I'm willing to merge.

If you write a document on command based programming for python, I'm willing to accept that as well.

File: commandbased/commandbasedrobot.py (hunk 1 above)
Comment (FRC2539) [inline commandbased/commandbasedrobot.py]: It wasn't obvious from the code that the two are equivalent. Here are their definitions.

```
def waitForData(self, timeout = None):
        with self.dataSem:
            self.dataSem.wait(timeout)
```

```
def isNewControlData(self):
        with self.mutex:
            result = self.newControlData
            self.newControlData = False
            return result
```

They seem to be doing very similar things, but they don't reference any of the same variables (although `dataSem` is created from `mutex`). When `waitForData()` returns is `isNewControlData()` guaranteed to be `True`. If so, we can remove the loop.

**Comment by FRC2539** [comment]
When you say we should make a PR to wpilib, do you mean robotpy-wpilib?

**Comment by virtuald** [comment]
I meant the C++/Java WPILib (in which case, we would include it in the RobotPy wpilib too). They're on github now: https://github.com/wpilibsuite/allwpilib .

File: commandbased/commandbasedrobot.py (hunk 1 above)
Comment (virtuald) [inline commandbased/commandbasedrobot.py]: Referring to [this line](https://github.com/robotpy/robotpy-wpilib/blob/master/wpilib/wpilib/driverstation.py#L113), I believe that guarantee can be made, as `getData` will always be called before the semaphore is notified.

It looks like they've updated their documentation in [wpilib](https://github.com/wpilibsuite/allwpilib/blob/0cd05d1a42aa7c08a212dfb40b00a28dea99efbe/wpilibc/athena/src/DriverStation.cpp#L441)... I'll file an issue to remind myself to update ours once I start on 2017 updates (November/December timeframe). 

I'll also ask for a documentation update to explicitly state this.... though, looking at their C++ implementation, it looks broken. >_>

**Comment by virtuald** [comment]
DS bugs filed at https://github.com/wpilibsuite/allwpilib/issues/233 and https://github.com/wpilibsuite/allwpilib/issues/234

**Comment by virtuald** [comment]
LGTM, thanks!

**Pull request merged**